runtime/derive/src/generate_asc_type.rs

```diff
@@ -0,0 +1,180 @@
+        "bool" => true,
+        _ => false,
+    }
+}
+
+fn field_type_map(tp: String) -> String {
+    if is_scalar(&tp) {
+        tp
+    } else {
+        match tp.as_ref() {
+            "String" => "graph_runtime_wasm::asc_abi::class::AscString".into(),
+            _ => tp.to_owned(),
+        }
+    }
+}
+
+fn field_type(fld: &syn::Field) -> String {
+    if let syn::Type::Path(tp) = &fld.ty {
+        if let Some(ps) = tp.path.segments.last() {
+            let name = ps.ident.to_string();
+            //TODO - this must be optimized
+            match name.as_ref() {
+                "Vec" => {
+                    //Vec<TxResult> -> AscPtr<AscTxResultArray>
+                    //Vec<u8>       - > AscPtr<Uint8Array>
+                    //::prost::alloc::vec::Vec<::prost::alloc::vec::Vec<u8>> -> AscPtr<AscBytesArray>,
+                    //::prost::alloc::vec::Vec<::prost::alloc::string::String> -> AscPtr<Array<AscPtr<AscString>>>,
+
+                    match &ps.arguments {
+                        syn::PathArguments::AngleBracketed(v) => {
+                            if let syn::GenericArgument::Type(syn::Type::Path(p)) = &v.args[0] {
+                                let nm = path_to_string(&p.path);
+
+                                match nm.as_ref(){
+                                    "u8" => "graph::runtime::AscPtr<graph_runtime_wasm::asc_abi::class::Uint8Array>".to_owned(),
+                                    //"Vec<u8>" => "graph::runtime::AscPtr<graph_runtime_wasm::asc_abi::class::AscBytesArray>".to_owned(),
+                                    "Vec<u8>" => "graph::runtime::AscPtr<crate::protobuf::AscBytesArray>".to_owned(),
+                                    "String" => "graph::runtime::AscPtr<crate::protobuf::Array<graph::runtime::AscPtr<graph_runtime_wasm::asc_abi::class::AscString>>>".to_owned(),
+                                    _ => format!("graph::runtime::AscPtr<crate::protobuf::Asc{}Array>", path_to_string(&p.path))
+                                }
+                            } else {
+                                name
+                            }
+                        }
+
+                        syn::PathArguments::None => name,
+                        syn::PathArguments::Parenthesized(_v) => {
+                            panic!("syn::PathArguments::Parenthesized is not implemented")
+                        }
+                    }
+                }
+                "Option" => match &ps.arguments {
+                    syn::PathArguments::AngleBracketed(v) => {
+                        if let syn::GenericArgument::Type(syn::Type::Path(p)) = &v.args[0] {
+                            let tp_nm = path_to_string(&p.path);
+                            if is_scalar(&tp_nm) {
+                                format!("Option<{}>", tp_nm)
+                            } else {
+                                format!("graph::runtime::AscPtr<crate::protobuf::Asc{}>", tp_nm)
```

Comment:
I imagine that's the reason for the introduction of the intermediary `protobuf` module. Should be documented in the top of the macro (each of them if it affects them all).

Reply:
there is https://crates.io/crates/protobuf (parser) and crate::protobuf
https://crates.io/crates/protobuf - is to parse proto files and annotate prost structs with meta info... unless we are missing your point... let's discuss